jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/service/WildFlyCustomRegionFactoryInitiator.java

```diff
@@ -38,14 +38,17 @@ protected RegionFactory resolveRegionFactory(Map configurationValues, ServiceReg
 
         // treat Hibernate 2lc as off, if not specified.
         // Note that Hibernate 2lc in 5.1.x, defaults to disabled, so this code is only needed in 5.3.x+.
-        if(Boolean.parseBoolean((String)use_second_level_cache) ||
-                (jpa_shared_code_mode != null || !"NONE".equals(jpa_shared_code_mode.toString()))) {
+        if(Boolean.parseBoolean((String)use_second_level_cache)) {
             configurationValues.put("hibernate.cache.region.factory_class", "org.infinispan.hibernate.cache.v53.InfinispanRegionFactory");
             return super.resolveRegionFactory(configurationValues, registry);
-        }
-        else {
+        } else if("UNSPECIFIED".equals(jpa_shared_code_mode.toString())
+             || "NONE".equals(jpa_shared_code_mode.toString())) {
             // explicitly disable 2lc cache
             return NoCachingRegionFactory.INSTANCE;
         }
+        else {
+            configurationValues.put("hibernate.cache.region.factory_class", "org.infinispan.hibernate.cache.v53.InfinispanRegionFactory");
+            return super.resolveRegionFactory(configurationValues, registry);
+        }
```

Comment:
This looks weird. Why don't we just use something like:
```
if("UNSPECIFIED".equals(jpa_shared_code_mode.toString())
     || "NONE".equals(jpa_shared_code_mode.toString())) {
    // explicitly disable 2lc cache
    return NoCachingRegionFactory.INSTANCE;
}
configurationValues.put("hibernate.cache.region.factory_class", "org.infinispan.hibernate.cache.v53.InfinispanRegionFactory");
return super.resolveRegionFactory(configurationValues, registry);
```

Also what is the rule for "UNSPECIFIED"? Is that value allowed in Java EE 7?

Reply:
Actually I see in the comment now that 5.1 defaults to off. The if statement is overly complex IMO though.

Reply:
If the Hibernate ORM use_second_level_cache configuration setting is specified by the application, that is our first choice.  

If the (JPA standard) jpa_shared_code_mode is specified as UNSPECIFIED, I current think we should disable the 2lc (which is what I did).  We might flip this later if possible, to default on.  If jpa_shared_code_mode is NONE, we will also disable 2lc.

Current code is:
```
        if(Boolean.parseBoolean((String)use_second_level_cache)) {
            configurationValues.put("hibernate.cache.region.factory_class", "org.infinispan.hibernate.cache.v53.InfinispanRegionFactory");
            return super.resolveRegionFactory(configurationValues, registry);
        } else if("UNSPECIFIED".equals(jpa_shared_code_mode.toString())
             || "NONE".equals(jpa_shared_code_mode.toString())) {
            // explicitly disable 2lc cache
            return NoCachingRegionFactory.INSTANCE;
        }
        else {
            configurationValues.put("hibernate.cache.region.factory_class", "org.infinispan.hibernate.cache.v53.InfinispanRegionFactory");
            return super.resolveRegionFactory(configurationValues, registry);
        }
```

Explanation of UNSPECIFIED from https://docs.oracle.com/javaee/7/api/javax/persistence/SharedCacheMode.html :
UNSPECIFIED - Caching behavior is undefined: provider-specific defaults may apply.
NONE - Caching is disabled for the persistence unit.